### PR TITLE
AP_BattMonitor: add PWM Fuel Level Sensor

### PR DESCRIPTION
**Pull request by tridge**
this sensor produces PWM between 1100 and 1900, giving fuel level as a
percentage
Driver written by James Pattison and myself

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+public:
+
+    /// Constructor
+    AP_BattMonitor_FuelLevel_PWM(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params);
+
+    /// Read the battery voltage and current.  Should be called at 10hz
+    void read() override;
+
+    /// returns true if battery monitor provides consumed energy info
+    bool has_consumed_energy() const override
```
Comment (rmackay9) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.h]: minor formatting thing but I would probably squash these down to one line.

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        // attach to new pin
+        last_pin = pin;
+        if (last_pin > 0) {
+            hal.gpio->pinMode(last_pin, HAL_GPIO_INPUT);
+            if (!hal.gpio->attach_interrupt(
+                    last_pin,
+                    FUNCTOR_BIND_MEMBER(&AP_BattMonitor_FuelLevel_PWM::irq_handler, void, uint8_t, bool, uint32_t),
+                    AP_HAL::GPIO::INTERRUPT_BOTH)) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "FuelLevelPWM: Failed to attach to pin %u", last_pin);
```
Comment (rmackay9) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: probably need to cast "last_pin" to "unsigned" or change the format to "%d" or we will end up with a compiler warning

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    /*
+      this driver assumes that CAPACITY is set to tank volume in millilitres.
+     */
+    const uint16_t pwm_empty = 1100;
+    const uint16_t pwm_full = 1900;
+    const uint16_t pwm_buffer = 20;
+
+
+    // check for invalid pulse
+    if (pulse_width <= (pwm_empty - pwm_buffer)|| pulse_width >= (pwm_full + pwm_buffer)) {
```
Comment (rmackay9) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: It's probably fine but it's a bit strict in requiring the pwm range to be between 1080 and 1920.  Personally I'd probably make it a1000~2000 or perhaps even 900~2100 if I was feeling generous.

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            if (!hal.gpio->attach_interrupt(
+                    last_pin,
+                    FUNCTOR_BIND_MEMBER(&AP_BattMonitor_FuelLevel_PWM::irq_handler, void, uint8_t, bool, uint32_t),
+                    AP_HAL::GPIO::INTERRUPT_BOTH)) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "FuelLevelPWM: Failed to attach to pin %u", last_pin);
+            }
+        }
+    }
+    uint32_t now_us = AP_HAL::micros();
+    if (pulse_count2 == irq_state.pulse_count1) {
```
Comment (rmackay9) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: We don't need to use the "hal.scheduler->disable_interrupts_save();" I guess?  I see the code is a little different than our AP_RSSI library's handling of the interrupts.

**Comment by rmackay9** [comment]
This will be great to have!  I know that at least one Partner has been using the AP_RSSI as a work around to having this.

The hardcoded range of 1100 to 1900 may be fine, or maybe not but I'm glad we're not immediately adding new parameters.

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: the device is defined as 1100 to 1900. We've added the margin to cope with timing. If another device comes along with a different range then we'll add the range to the backend constructor and add a new enum for the type

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: no, I don't think we need to as we are not accumulating, its a point sample

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.h]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelLevel_PWM.cpp]: done

**Pull request merged**